File: String/isPalindromeExpanded/isPalindromeExpanded.cpp
#include <iostream>
#include <string>
#include <cctype>

bool is_palindrome(const std::string &str) {
	unsigned b = 0;
	unsigned e = str.size() - 1;
	while (b < e) {
		while (!isalpha(str[b]) && b < e) ++b;
		while (!isalpha(str[e]) && b < e)--e;
		if (tolower(str[b]) != tolower(str[e])) return false;
		++b; --e;
	}
	
	return true;
}

int main() {
	std::cout << "Is \"racecar\" a palindrome: " 
				<< (is_palindrome("racecar") ? "true":"false") 
				<< std::endl; 

	std::cout << "Is \"Eva, Can I Stab Bats In A Cave?\" a palindrome: " 
				<< (is_palindrome("	Eva, Can I Stab Bats In A Cave?") ? "true":"false") 
				<< std::endl;
				
	std::cout << "Is \"Not a palindrome\" a palindrome: " 
				<< (is_palindrome("Not a palindrome") ? "true":"false") 
				<< std::endl; 
				
	// If we consider only alphabetic characters, then this is palindrome
	// since 'racecar' is the only thing here.
	std::cout << "Is \"r131221a353c31234//326e001-car\" a palindrome: " 
				<< (is_palindrome("r131221a353c31234//326e001-car") ? "true":"false") 
				<< std::endl; 
}

File: String/isPalindromeExpanded/isPalindromeExpanded.go
package main

import (
	"bytes"
	"fmt"
	"unicode"
)

func isPalindromeExpanded(str string) bool {
	runes := []rune(str)
	front := 0
	back := len(runes) - 1
	for front < back {
		for !unicode.IsLetter(runes[front]) {
			front++
		}
		for !unicode.IsLetter(runes[back]) {
			back--
		}
		if unicode.ToLower(runes[front]) != unicode.ToLower(runes[back]) {
			return false
		}
		front++
		back--
	}
	return true
}

func main() {
	fmt.Printf(isPalindromePrint("racecar"))                          // true
	fmt.Printf(isPalindromePrint("	Eva, Can I Stab Bats In A Cave?")) // true
	fmt.Printf(isPalindromePrint("Not a palindrome"))                 // false
	fmt.Printf(isPalindromePrint("r131221a353c31234//326e001-car"))   // true
}

func isPalindromePrint(str string) string {
	var buffer bytes.Buffer
	buffer.WriteString(fmt.Sprintf("Is \"%s\" a palindrome: ", str))
	if isPalindromeExpanded(str) {
		buffer.WriteString("true")
	} else {
		buffer.WriteString("false")
	}
	buffer.WriteString("\n")
	return buffer.String()
}
